crates/tx5/src/ep3.rs

```diff
@@ -0,0 +1,600 @@
+//! Module containing tx5 endpoint version 3 types.
+
+use crate::deps::lair_keystore_api;
+use crate::deps::sodoken;
+use crate::AbortableTimedSharedFuture;
+use crate::BackBuf;
+use futures::future::BoxFuture;
+use lair_keystore_api::prelude::*;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex, Weak};
+use tx5_core::{Error, EventRecv, EventSend, Id, Result, Tx5Url};
+
+fn next_uniq() -> u64 {
+    static UNIQ: std::sync::atomic::AtomicU64 =
+        std::sync::atomic::AtomicU64::new(1);
+    UNIQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+}
+
+type CRes<T> = std::result::Result<T, Error>;
+
+/// Events generated by a tx5 endpoint version 3.
+pub enum Ep3Event {
+    /// An fatal error indicating the endpoint is no longer viable.
+    Error(Error),
+
+    /// Connection established.
+    Connected {
+        /// Url of the remote peer.
+        peer_url: PeerUrl,
+    },
+
+    /// Connection closed.
+    Disconnected {
+        /// Url of the remote peer.
+        peer_url: PeerUrl,
+    },
+
+    /// Receiving an incoming message from a remote peer.
+    Message {
+        /// Url of the remote peer.
+        peer_url: PeerUrl,
+
+        /// Message sent by the remote peer.
+        message: Vec<u8>,
+
+        /// Permit counting the bytes allowed in memory on the receive side.
+        permit: tokio::sync::OwnedSemaphorePermit,
+    },
+}
+
+impl From<Error> for Ep3Event {
+    fn from(err: Error) -> Self {
+        Self::Error(err)
+    }
+}
+
+impl std::fmt::Debug for Ep3Event {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Error(err) => {
+                f.debug_struct("Error").field("err", err).finish()
+            }
+            Self::Connected { peer_url } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Connected").field("peer_url", &url).finish()
+            }
+            Self::Disconnected { peer_url } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Disconnected")
+                    .field("peer_url", &url)
+                    .finish()
+            }
+            Self::Message { peer_url, .. } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Message").field("peer_url", &url).finish()
+            }
+        }
+    }
+}
+
+/// A signal server url.
+pub type SigUrl = Tx5Url;
+
+/// A peer connection url.
+pub type PeerUrl = Tx5Url;
+
+type SigMap = HashMap<SigUrl, (u64, AbortableTimedSharedFuture<Arc<Sig>>)>;
```

Comment:
This `AbortableTimedSharedFuture` is the core enabling mechanic of this refactor, you can find the definition in `crates/tx5/src/lib.rs`. The first thread/task to ask for a particular signal server will add an entry to this hash map, then all future requests will await on that same future. The complexity comes in cleaning these up in the case of errors.